docs/MV3.md

```diff
@@ -0,0 +1,114 @@
+
+## Implementation
+
+The most important change that lead to this migration was how url request interception model was changed. In MV2, we used `webRequest.onBeforeRequest` to intercept requests and redirect them to the local gateway. That process looked something like:
+
+![MV2 Working Diagram](assets/mv2-working.png)
+
+The process was simple and synchronous, the request is intercepted and handed over to companion, which then redirects it to the local gateway as needed. MV3 changes this process by introducing a new API called `declarativeNetRequest`. MV3 does not allow synchronous request interception (also called blocking request), which is not secure and can lead to performance issues. Instead, it uses a declarative approach where the extension declares a set of rules that are then used by the browser to intercept requests. This process looks something like:
+
+![MV3 Working Diagram](assets/mv3-working.png)
+
+The process is asynchronous, the browser allows "observation" of requests to companion, which asynchronously determines if the given request is serviceable by IPFS and then dynamically introduces a rule for the browser to perform redirects to the local gateway. A sample rule looks something like:
+
+```js
+{
+  id,                                    // random id for reference
+  priority: 1,                           // priority of the rule
+  action: {
+    type: 'redirect',
+    redirect: { '<regex substitution>' } // redirect to gateway
+  },
+  condition: {
+    '<regex filter>',                    // filter to match requests
+    '<filtered domains list>,
+    resourceTypes: [
+      'csp_report',
+      'font',
+      'image',
+      'main_frame',
+      'media',
+      'object',
+      'other',
+      'ping',
+      'script',
+      'stylesheet',
+      'sub_frame',
+      'webbundle',
+      'websocket',
+      'webtransport',
+      'xmlhttprequest'
+    ]
+  }
+}
+```
+
+The API only allows for 5000 such rules to exist at the time of writing this, so that value can be exhausted pretty quickly. So instead of adding one rule per redirect, we can dynamically generate a single regex pattern to handle multiple redirections.
+
+For example
+
+```
+https://ipfs.io/ipns/en.wikipedia-on-ipfs.org -> http://localhost:8080/ipns/en.wikipedia-on-ipfs.org
+```
+
+can be represented as
+
+```
+https://ipfs.io/ipns/(.*) -> http://localhost:8080/ipns/$1
+```
+
+That's just an example, but in reality it looks more like (which covers edge cases like URL endings):
+
+```js
+{
+  "action": {
+    "redirect": {
+      "regexSubstitution": "http://localhost:8080\\1"
+    },
+    "type": "redirect"
+  },
+  "condition": {
+    "excludedInitiatorDomains": [],
+    "regexFilter": "^https?\\:\\/\\/ipfs\\.io((?:[^\\.]|$).*)$",
```

Comment:
Above we talked about "/ipns" urls.  Here our regex passes everything under ipfs.io not just everything under ipfs.io/ipns.  Is that intentional?